### PR TITLE
Add variable length functionality to DataLoaders

### PR DESCRIPTION
**Pull request by canerturkmen**
*Issue #, if available:* Related to #294 

*Description of changes:*

Currently data loaders assume that arrays (e.g., that come out of a splitter type transform) are easily stackable into data batches. This is not convenient when working with time series of variable length that a model may want to see, such as when working with temporal point processes.

Here, we add variable-length awareness to the `batchify` and `stack` functions in the parallelized loader. We also expose the batchify function as a constructor argument to `ParallelDataLoader`. A consumer of DataLoaders can now specify `DataLoader(..., variable_length=True)` in order to make the loader aware that some input fields may not be stackable. In this case, the loader pads the arrays on the right along the first dimension (this is a hardcoded assumption for now) before stacking them. By default, the loader falls back to business as usual. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by canerturkmen** [comment]
Test fails are likely related to the GPU. @AaronSpieler  can you maybe take a look when you have time? Tests run fine locally.

**Comment by AaronSpieler** [comment]
> Test fails are likely related to the GPU. @AaronSpieler can you maybe take a look when you have time? Tests run fine locally.

So MXNet cant handle pickling empty arrays, could this be the problem?

**Comment by AaronSpieler** [comment]
In general: I'm not sure whether this is the place to do this. As our codebase currently is, I would suspects its the `InstanceSplitter`, since it has also functionality like `pick_incomplete` in which case time series are also padded. Putting this code here leads to a preprocessing step happening outside the Transformation Chain.

I would suggest to consider for your use case using the `InstanceSplitter`: https://gluon-ts.mxnet.io/api/gluonts/gluonts.transform.html?highlight=instancesplitter#gluonts.transform.InstanceSplitter and 
1) extending the functionality you need there if it't not there. 
2) Then, defining a transformation chain with the splitter
The Transformation Chain is always executed in the DataLoader.

What do you think @canerturkmen ?

**Comment by canerturkmen** [comment]
> > Test fails are likely related to the GPU. @AaronSpieler can you maybe take a look when you have time? Tests run fine locally.
> 
> So MXNet cant handle pickling empty arrays, could this be the problem?

Yes this appears to be the case, ends up in (I think) a segfault. Bad thing is, this is a valid edge case.

**Comment by canerturkmen** [comment]
> In general: I'm not sure whether this is the place to do this. As our codebase currently is, I would suspects its the `InstanceSplitter`, since it has also functionality like `pick_incomplete` in which case time series are also padded. Putting this code here leads to a preprocessing step happening outside the Transformation Chain.
> 
> I would suggest to consider for your use case using the `InstanceSplitter`: https://gluon-ts.mxnet.io/api/gluonts/gluonts.transform.html?highlight=instancesplitter#gluonts.transform.InstanceSplitter and
> 
>     1. extending the functionality you need there if it't not there.
> 
>     2. Then, defining a transformation chain with the splitter
>        The Transformation Chain is always executed in the DataLoader.
> 
> 
> What do you think @canerturkmen ?

So this was tricky in the original contribution as well. Unfortunately, the padding is a function of the batch itself, which materializes only through the data loader. So cramming it into a transformer would, in my mind, be the same as introducing a "batch transformer," or imitate "batching" in a transformer. Our transformers are designed to operate on logical `DataEntry`s so I think we should have it in the loader.

This is a good discussion though, thanks for bringing this up @AaronSpieler . What do you think @lostella?

**Comment by AaronSpieler** [comment]
> > > Test fails are likely related to the GPU. @AaronSpieler can you maybe take a look when you have time? Tests run fine locally.
> > 
> > 
> > So MXNet cant handle pickling empty arrays, could this be the problem?
> 
> Yes this appears to be the case, ends up in (I think) a segfault. Bad thing is, this is a valid edge case.

I ended up converting those matrices to `numpy`, but in your use-case this probably doesn't work.

**Comment by AaronSpieler** [comment]
> "batch transformer," or imitate "batching"

Well, wouldn't then the current `InstanceSplitter` already be a  "batch transformer" since it's padding sequences to a common length, which is exactly your use case as I understand?

**Comment by canerturkmen** [comment]
1. For the 0-length issue, I'll probably just have a dummy array and handle it downstream. Ugly, but necessary and should be easy to handle.
2. On the splitter issue. Yes, but two key things are different here:
   - the `InstanceSplitter` pads for the *entire data set,* here pad length is a function of the *batch* itself. 
   - the `InstanceSplitter` knows the lengths of arrays ahead of time (indeed, it's a parameter). Here the lengths are variable (from the `ContinuousTimeInstanceSplitter`).

A potential workaround is to have a PadEverythingSplitter which would be wasteful in two ways.
- It would need to fully scan the data set to know the lengths of everyone, which goes against the spirit of our Datasets. I understand from the current design that transformers should not be picky about which type of dataset they work on (so, not necessarily a `ListDataset`), and be able to handle iterators. A PadEverythingSplitter maybe would have to exhaust the data set to decide which length to pad to.
- It could be potentially very wasteful with memory, up to 2x, 3x more footprint. Given the GPU constraint this blocks other nice things downstream. Our TPPs will be multivariate, so very demanding with memory (and hence the initial thinking).

**Comment by AaronSpieler** [comment]
> 1. For the 0-length issue, I'll probably just have a dummy array and handle it downstream. Ugly, but necessary and should be easy to handle.
> 2. On the splitter issue. Yes, but two key things are different here:
>    
>    * the `InstanceSplitter` pads for the _entire data set,_ here pad length is a function of the _batch_ itself.
>    * the `InstanceSplitter` knows the lengths of arrays ahead of time (indeed, it's a parameter). Here the lengths are variable (from the `ContinuousTimeInstanceSplitter`).
> 
> A potential workaround is to have a `PadEverythingTransformer` which would be wasteful in two ways.
> 
> * It would need to fully scan the data set to know the lengths of everyone, which goes against the spirit of our Datasets. I understand from the current design that transformers should not be picky about which type of dataset they work on (so, not necessarily a `ListDataset`), and be able to handle iterators. A `PadEverythingTransformer` maybe would have to exhaust the data set to decide which length to pad to.
> * It could be potentially very wasteful with memory, up to 2x, 3x more footprint. Given the GPU constraint this blocks other nice things downstream. Our TPPs will be multivariate, so very demanding with memory (and hence the initial thinking).
 
* Regarding 1: yeah, probably the easiest fix

* Regarding 2: 
Does this mean that individual batches can have different length of dimension 1:

"the InstanceSplitter pads for the entire data set, here pad length is a function of the batch itself."

How would setting a `max_length` affect the implementation, and would it be sensible? So to say to always pad to this length, even if the longest time series in the batch is not that long.

**Comment by canerturkmen** [comment]
@AaronSpieler Yes. Individual batches can have different lengths. Unfortunately, `max_length` in this case is not really an option both on account of the memory constraint and more importantly it is near-impossible to set intuitively.

**Comment by AaronSpieler** [comment]
> @AaronSpieler Yes. Individual batches can have different lengths. Unfortunately, `max_length` in this case is not really an option both on account of the memory constraint and more importantly it is near-impossible to set intuitively.

I see, Thx for the clarification!

File: src/gluonts/dataset/parallelized_loader.py [hunk 1]
```diff
@@ -84,16 +85,77 @@ def reduce_ndarray(data):
 ForkingPickler.register(nd.NDArray, reduce_ndarray)
 
 
+def _is_stackable(
+    arrays: List[Union[np.ndarray, mx.nd.NDArray, Any]],
```
Comment (AaronSpieler) [inline src/gluonts/dataset/parallelized_loader.py:89]: Is there any point in having this Union, if it contains Any?

File: src/gluonts/dataset/parallelized_loader.py [hunk 2]
```diff
@@ -84,16 +85,77 @@ def reduce_ndarray(data):
...
+    assert isinstance(data[0], (np.ndarray, mx.nd.NDArray))
+    is_mx = isinstance(data[0], mx.nd.NDArray)
+
+    # MxNet causes a segfault when persisting 0-length arrays. As such,
+    # we add a dummy pad of length 1 to 0-length dims.
+    max_len = max(1, functools.reduce(max, (x.shape[pad_axis] for x in data)))
+    padded_data = []
+
+    for x in data:
+        # MxNet lacks the functionality to pad n-D arrays consistently.
```
Comment (AaronSpieler) [inline src/gluonts/dataset/parallelized_loader.py:113]: So `mxnet.ndarray.pad`: https://beta.mxnet.io/api/ndarray/_autogen/mxnet.ndarray.pad.html is not an option here?

File: src/gluonts/dataset/parallelized_loader.py [hunk 3]
```diff
@@ -84,16 +85,77 @@ def reduce_ndarray(data):
...
+    return True
+
+
+def _pad_arrays(
+    data: List[Union[np.ndarray, mx.nd.NDArray]], pad_axis: int = 0,
+) -> List[Union[np.ndarray, mx.nd.NDArray]]:
+    assert isinstance(data[0], (np.ndarray, mx.nd.NDArray))
+    is_mx = isinstance(data[0], mx.nd.NDArray)
+
+    # MxNet causes a segfault when persisting 0-length arrays. As such,
```
Comment (AaronSpieler) [inline src/gluonts/dataset/parallelized_loader.py:107]: So I'm wondering, why does it come to this, i.e. why would we want to persist 0-length arrays? Can we not just pad the whole length with 0s or whatever the padding value is? 
What about removing that entry from the batch?

File: src/gluonts/dataset/parallelized_loader.py [hunk 4]
```diff
@@ -84,16 +85,77 @@ def reduce_ndarray(data):
...
+    # we add a dummy pad of length 1 to 0-length dims.
+    max_len = max(1, functools.reduce(max, (x.shape[pad_axis] for x in data)))
+    padded_data = []
+
+    for x in data:
+        # MxNet lacks the functionality to pad n-D arrays consistently.
+        # We fall back to numpy if x is an mx.nd.NDArray.
```
Comment (AaronSpieler) [inline src/gluonts/dataset/parallelized_loader.py:114]: So this could mean a lot of overhead, we should be extra sure that there is no way of padding MXNet arrays properly.

File: src/gluonts/dataset/parallelized_loader.py (hunk 1 above)
Comment (canerturkmen) [reply src/gluonts/dataset/parallelized_loader.py:89]: If I put `int`, `float` instead of `Any`, mypy complains despite the isinstance() check below. So, those are just meant to be "type hints" I guess.

File: src/gluonts/dataset/parallelized_loader.py (hunk 2 above)
Comment (canerturkmen) [reply src/gluonts/dataset/parallelized_loader.py:113]: From the docs:
```
Current implementation only supports 4D and 5D input arrays with padding applied only on axes 1, 2 and 3. Expects axes 4 and 5 in pad_width to be zero.
```
🤷‍♂️

File: src/gluonts/dataset/parallelized_loader.py (hunk 3 above)
Comment (canerturkmen) [reply src/gluonts/dataset/parallelized_loader.py:107]: Persisting is not really relevant here. Under normal circumstances we do pad 0-length arrays to the appropriate length. This line only kicks in when the *entire batch* has 0-length, in which case we pad to a dummy dimensionality of 1.

Unfortunately the entry is valid data. You want to know if you observed nothing--can't kick it out of the batch.

File: src/gluonts/dataset/parallelized_loader.py (hunk 2 above)
Comment (AaronSpieler) [reply src/gluonts/dataset/parallelized_loader.py:113]: Ok, super weird.

File: src/gluonts/dataset/parallelized_loader.py (hunk 3 above)
Comment (AaronSpieler) [reply src/gluonts/dataset/parallelized_loader.py:107]: I see.

**Comment by vafl** [comment]
Perhaps we can expose the batch / stack function as a parameter of the Trainer and pass it along? Then you could plug in whatever you want. For inference if we use batch_size=1 it should work without changes, or we can also pass this to a predictor somehow.

File: src/gluonts/dataset/parallelized_loader.py (hunk 2 above)
Comment (AaronSpieler) [reply src/gluonts/dataset/parallelized_loader.py:113]: Can you please test whether the workaround described here: https://mxnet.apache.org/versions/1.4.1/tutorials/gluon/gotchas_numpy_in_mxnet.html
makes sense?

**Comment by canerturkmen** [comment]
> Perhaps we can expose the batch / stack function as a parameter of the Trainer and pass it along? Then you could plug in whatever you want. For inference if we use batch_size=1 it should work without changes, or we can also pass this to a predictor somehow.

Thanks for popping in @vafl. I'm not sure I fully understand but I'll try to comment nevertheless :) Both `batchify` and `stack` make sense to me in the context of loaders, and best remain within the loader module boundaries.  With this change, the loader exposes the batching function as a constructor argument,
```python
TrainDataLoader(
                num_batches_per_epoch=22, 
                batchify_fn=partial(batchify, variable_length=True), 
                **kwargs
            )
```
I understand our Trainers do not know about TrainDataLoaders (don't keep references to them in their state), and only get them as call arguments. It's the Estimator that calls the loader's constructor, and passes it to the trainer. Do you think Estimators should expose batchify functions?

**Comment by vafl** [comment]
> > Perhaps we can expose the batch / stack function as a parameter of the Trainer and pass it along? Then you could plug in whatever you want. For inference if we use batch_size=1 it should work without changes, or we can also pass this to a predictor somehow.
> 
> Thanks for popping in @vafl. I'm not sure I fully understand but I'll try to comment nevertheless :) Both `batchify` and `stack` make sense to me in the context of loaders, and best remain within the loader module boundaries. With this change, the loader exposes the batching function as a constructor argument,
> 
> ```python
> TrainDataLoader(
>                 num_batches_per_epoch=22, 
>                 batchify_fn=partial(batchify, variable_length=True), 
>                 **kwargs
>             )
> ```
> 
> I understand our Trainers do not know about TrainDataLoaders (don't keep references to them in their state), and only get them as call arguments. It's the Estimator that calls the loader's constructor, and passes it to the trainer. Do you think Estimators should expose batchify functions?

Sorry, yeah. The trainer is not the right place.

What I meant in general was whether we can make this more flexible and configurable from the outside. So for example, we could make `stack_fn` a parameter of `batchify` with the default being the standard `stack` and then you could tie that to another stack function in your estimator. This is just an idea. I think your change also makes sense as is.

File: src/gluonts/dataset/parallelized_loader.py (hunk 2 above)
Comment (canerturkmen) [reply src/gluonts/dataset/parallelized_loader.py]: So I implemented this, see latest commit `c10ded2`.

Tests pass locally, but still fail with a segfault (?). If you also can't spot the problem in the code, I suggest we go back to the old solution. This should only be used by TPPs for a long time, and we can try to work around it.

File: src/gluonts/dataset/parallelized_loader.py [hunk 5]
```diff
@@ -84,16 +85,77 @@ def reduce_ndarray(data):
 ForkingPickler.register(nd.NDArray, reduce_ndarray)
 
 
+def _is_stackable(
+    arrays: List[Union[np.ndarray, mx.nd.NDArray, Any]],
+) -> bool:
+    """
+    Check if elements are scalars, have too few dimensions, or their
+    time axes have equal length; i.e. they are directly `stack` able.
```
Comment (lostella) [inline src/gluonts/dataset/parallelized_loader.py:93]: Just to align this with the padding function that follows: the axis being checked could be an (optional) argument

File: src/gluonts/dataset/parallelized_loader.py [hunk 6]
```diff
@@ -84,16 +85,77 @@ def reduce_ndarray(data):
...
+    """
+    if isinstance(arrays[0], (mx.nd.NDArray, np.ndarray)):
+        s = set(arr.shape[0] for arr in arrays)
+        return len(s) <= 1 and arrays[0].shape[0] != 0
+    return True
+
+
+def _pad_arrays(
+    data: List[Union[np.ndarray, mx.nd.NDArray]], pad_axis: int = 0,
```
Comment (lostella) [inline src/gluonts/dataset/parallelized_loader.py:102]: Maybe just “axis” instead of “pad_axis”: more like mxnet B-) and less redundant

File: src/gluonts/dataset/parallelized_loader.py (hunk 2 above)
Comment (AaronSpieler) [reply src/gluonts/dataset/parallelized_loader.py:113]: I see, then we'll have to look into it some other time. Anyways, thx for trying!
